### PR TITLE
Goblins!

### PR DESCRIPTION
**Pull request by nc6**

### CODE REVIEW

**Comment by mdimjasevic** [comment]
@nc6, can you provide a brief summary of this pull request?

**Comment by commandodev** [comment]
@nc6 looks like we have some conflicts now. Other than that is everyone happy to merge it?

File: specs/semantics/hs/src/Control/State/Transition/Generator.hs [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+  } deriving Generic
+
+-- | Generation monad
+newtype GenM s a = GenM (GenT (MS.State (Bookkeeping s)) a)
+  deriving (Functor, Applicative, Monad)
+  deriving (HasState "bkGenState" (GenState s))
+    via Field "bkGenState" () (MonadState (GenT (MS.State (Bookkeeping s))))
+
+deriving via Field "bkState" () (MonadState (GenT (MS.State (Bookkeeping s))))
+  instance (st ~ State s) => HasState "bkState" st (GenM s)
```
Comment (dnadales) [inline specs/semantics/hs/src/Control/State/Transition/Generator.hs:63]: One noob question Nick, why do you need these tilde (`~`) constraints here?

File: specs/semantics/hs/src/Control/State/Transition/Generator.hs [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+data Bookkeeping s = Bookkeeping
+  { bkGenState    :: GenState s
+  , bkState       :: State s
+  , bkEnvironment :: Environment s
+  , bkTrace       :: [(Signal s, [PredicateFailure s])]
+  } deriving Generic
+
+-- | Generation monad
+newtype GenM s a = GenM (GenT (MS.State (Bookkeeping s)) a)
```
Comment (dnadales) [inline specs/semantics/hs/src/Control/State/Transition/Generator.hs:57]: We use hedgehog's `GenT` but last time I tried I couldn't manage to extract it without resorting to the `Internals` module: https://github.com/hedgehogqa/haskell-hedgehog/issues/208

File: specs/semantics/hs/src/Control/State/Transition/Generator.hs (hunk 1 above)
Comment (nc6) [reply specs/semantics/hs/src/Control/State/Transition/Generator.hs:63]: It relates to something to do with GHC handles unification, I think. But I can't recall the exact issue. I'll try to look it up and document!

**Pull request closed without merging**